components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +69,150 @@ impl MutationObserver {
+        MutationObserver::queueMutationObserverCompoundMicrotask(compoundMicrotask);
+    }
+}
+
+impl MutationObserverMethods for MutationObserver {
+    /// https://dom.spec.whatwg.org/#dom-mutationobserver-observe
+    /// MutationObserver.observe method
+    fn Observe(&self, target: &Node, options: &MutationObserverInit) -> Fallible<()> {
+        // Step 1: If either options’ attributeOldValue or attributeFilter is present and
+        // options’ attributes is omitted, set options’ attributes to true.
+        if (options.attributeOldValue.is_some() || options.attributeFilter.is_some()) && options.attributes.is_none() {
+//            options.attributes = Some(true);
+        }
+        // Step2: If options’ characterDataOldValue is present and options’ characterData is omitted,
+        // set options’ characterData to true.
+        if options.characterDataOldValue.is_some() && options.characterData.is_some() {
+//            options.characterData = Some(true);
+        }
+        // Step3: If none of options’ childList, attributes, and characterData is true, throw a TypeError.
+        if !options.childList && !options.attributes.unwrap() && !options.characterData.unwrap() {
+            return Err(Error::Type("childList, attributes, and characterData not true".to_owned()));
+        }
+        // Step4: If options’ attributeOldValue is true and options’ attributes is false, throw a TypeError.
+        if options.attributeOldValue.unwrap() && !options.attributes.unwrap() {
+            return Err(Error::Type("attributeOldValue is true but attributes is false".to_owned()));
+        }
+        // Step5: If options’ attributeFilter is present and options’ attributes is false, throw a TypeError.
+        if options.attributeFilter.is_some() && !options.attributes.unwrap() {
+            return Err(Error::Type("attributeFilter is present but attributes is false".to_owned()));
+        }
+        // Step6: If options’ characterDataOldValue is true and options’ characterData is false, throw a TypeError.
+        if options.characterDataOldValue.unwrap() && !options.characterData.unwrap() {
+            return Err(Error::Type("characterDataOldValue is true but characterData is false".to_owned()));
+        }
+        // TODO: Step 7
+        for registered in target.registered_mutation_observers_for_type().borrow().iter() {
+            // registered observer in target’s list of registered observers whose observer is the context object
+            if &*registered as *const MutationObserver == self as *const MutationObserver {
+                // TODO: remove matching transient registered observers
+                registered.options = Cell::from(*options);
+            }
+            // TODO: Step 8
+            let callback: Rc<MutationCallback>;
+            let observer: MutationObserver;
+            observer.options = Cell::from(*options);
+            target.add_registered_mutation_observer(observer);
```

Comment:
Shouldn't be we adding `self` here? However, we should only add a new observer if no matching one was found in the list (hence `otherwise` in step 8).